core/rpc/src/chain/number.rs

```diff
@@ -40,7 +40,14 @@ impl<Number: traits::As<u64>> NumberOrHex<Number> {
 	/// Fails in case hex number is too big.
 	pub fn to_number(self) -> Result<Number, String> {
 		match self {
-			NumberOrHex::Number(n) => Ok(n),
+			NumberOrHex::Number(n) => {
```

Comment:
Let's rewrite this to something like that:

```
let num: u64 = match self {
			NumberOrHex::Number(n) => n.as_(),
			NumberOrHex::Hex(h) => {
				// FIXME #1377 this only supports `u64` since `BlockNumber`
				// is `As<u64>` we could possibly go with `u128`.
				let l = h.low_u64();
				if U256::from(l) != h {
					return Err(format!("`{}` does not fit into the block number type.", h));
				} else {
				    l
				}
			}
		};

// FIXME <issuenumber>: Database seems to limit the block number to u32 for no reason
if num > u32::max_value() as u64 {
        Err(format!("`{}` > u32::max_value(), the max block number is u32.", num))
} else {
		Ok(traits::As::sa(num))
}
```